src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
+                     PROCESS_INFORMATION* const piPty,                  // _Out_
+                     const EnvironmentVariableMapW& extraEnvVars = {}); // _In_
 
 bool SignalResizeWindow(const HANDLE hSignal,
                         const unsigned short w,
                         const unsigned short h);
 
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map);
+
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars);
+
+// Function Description:
+// - Updates an EnvironmentVariableMapW with the current process's unicode
+//   environment variables ignoring ones already set in the provided map.
+// Arguments:
+// - map: The map to populate with the current processes's environment variables.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map)
+{
+    LPWCH currentEnvVars{};
+    auto freeCurrentEnv = wil::scope_exit([&] {
+        if (currentEnvVars)
+        {
+            (void)FreeEnvironmentStringsW(currentEnvVars);
+            currentEnvVars = nullptr;
+        }
+    });
+
+    RETURN_IF_NULL_ALLOC(currentEnvVars = ::GetEnvironmentStringsW());
+
+    // Block is guaranteed to be double-NULL terminated at a minimum.
+    for (wchar_t const* lastCh{ currentEnvVars }; *lastCh != '\0';)
+    {
+        // Copy current entry into temporary map.
+        const size_t cchEntry{ ::wcslen(lastCh) };
+        const std::wstring_view entry{ lastCh, cchEntry };
+
+        // Every entry is of the form "name=value\0".
+        auto pos = entry.find_first_of(L"=", 0, 1);
+        RETURN_HR_IF(E_UNEXPECTED, pos == std::wstring::npos);
+
+        std::wstring name{ lastCh, pos };
+        std::wstring value{ lastCh + pos + 1, cchEntry - pos - 1 };
+
+        // Don't replace entries that already exist.
+        map.try_emplace(std::move(name), std::move(value));
+
+        lastCh += cchEntry + 1; // Each entry is NULL-terminated.
+    }
+
+    return S_OK;
+}
+
+// Function Description:
+// - Creates a new environment block using the provided vector as appropriate
+//   (resizing if needed) based on the provided environment variable map
+//   matching the format of GetEnvironmentStringsW.
+// Arguments:
+// - map: The map to populate the new environment block vector with.
+// - newEnvVars: The vector that will be used to create the new environment block.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars)
+{
+    // Clear environment block before use.
+    constexpr size_t cbChar{ sizeof(decltype(newEnvVars.begin())::value_type) };
+
+    if (!newEnvVars.empty())
+    {
+        ::SecureZeroMemory(newEnvVars.data(), newEnvVars.size() * cbChar);

```

Comment:
Why not call `vector::clear()` here?

Reply:
clear() just runs the destructor on all elements in the vector and then sets the last pointer to the first pointer--it doesn't memset (zero) the actual memory.

So the call to [SecureZeroMemory](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/wdm/nf-wdm-rtlsecurezeromemory) is necessary to ensure I've actually cleared this data from the stack / heap.